### PR TITLE
[SPARK-32148][SS] Fix stream-stream join issue on missing to copy reused unsafe row

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This patch fixes the odd join result being occurred from stream-stream join for state store format V2.

There're some spots on V2 path which leverage UnsafeProjection. As the result row is reused, the row should be copied to avoid changing value during reading (or make sure the caller doesn't affect by such behavior) but `SymmetricHashJoinStateManager.removeByValueCondition` violates the case.

This patch makes `KeyWithIndexToValueRowConverterV2.convertValue` copy the row by itself so that callers don't need to take care about it. This patch doesn't change the behavior of `KeyWithIndexToValueRowConverterV2.convertToValueRow` to avoid double-copying, as the caller is expected to store the row which the implementation of state store will call `copy()`.

This patch adds such behavior into each method doc in `KeyWithIndexToValueRowConverter`, so that further contributors can read through and make sure the change / new addition doesn't break the contract.

### Why are the changes needed?

Stream-stream join with state store format V2 (newly added in Spark 3.0.0) has a serious correctness bug which brings indeterministic result.

### Does this PR introduce _any_ user-facing change?

Yes, some of Spark 3.0.0 users using stream-stream join from the new checkpoint (as the bug exists to only v2 format path) may encounter wrong join result. This patch will fix it.

### How was this patch tested?

Reported case is converted to the new UT, and confirmed UT passed. All UTs in StreamingInnerJoinSuite and StreamingOuterJoinSuite passed as well

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
It's reported as left-outer join specific, but given the possibility of impact I took the representation of "stream-stream join" instead of "left/right outer stream-stream join".

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing @jose-torres @gaborgsomogyi @xuanyuanking @uncleGen

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala [hunk 1]
```diff
@@ -295,6 +295,10 @@ case class StreamingSymmetricHashJoinExec(
             postJoinFilter(joinedRow.withLeft(leftKeyValue.value).withRight(rightValue))
           }
         }
+
+        // NOTE: we need to make sure `outerOutputIter` is evaluated "after" exhausting all of
+        // elements in `innerOutputIter`, because evaluation of `innerOutputIter` may update
+        // the match flag which the logic for outer join is relying on.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala:301]: Just to clarify: this comment is not related to the bug and just to document an existing assumption?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala [hunk 2]
```diff
@@ -451,10 +451,25 @@ class SymmetricHashJoinStateManager(
...
 
   private trait KeyWithIndexToValueRowConverter {
+    /** Defines the schema of the value row (the value side of K-V in state store). */
     def valueAttributes: Seq[Attribute]
 
+    /**
+     * Convert the value row to (actual value, match) pair.
+     *
+     * NOTE: implementations should ensure the result row is NOT reused during execution, as
+     * caller may use the value to store without copy().
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:461]: I don't know the details, but how likely the caller side will store the values?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala:301]: Yes right.

TBH I suspected this first and crafted a patch including the part with new iterator explicitly runs the logic after evaluating innerOutputIter, and later realized current logic already dealt with this properly, because removeOldState() doesn't materialize the candidates and evaluate lazily. This patch contains minimal change.

Worth to mention how it works for someone who may need to touch here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:461]: So the problem occurred here:

https://github.com/apache/spark/blob/42f01e314b4874236544cc8b94bef766269385ee/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala#L252-L281

Here we temporarily store `currentValue` via calling `convertValue` which is candidate to remove. After that we adjust the array a bit to not having a hole in between, where it also triggers `convertValue` and "changes" the value in `currentValue` silently if we don't copy.

It's also possible for caller to call copy() but the row to copy is inside of ValueAndMatchPair so I feel it's pretty much easier to miss even with method doc, so I took safer approach rather than performance-wise.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:461]: I don't have a strong opinion here. Just for reference: The SMJ copies the input row on-demand: only copy the rows when we need to buffer them. See `SortMergeJoinExec.bufferMatchingRows`. It's only a few places we need to buffer the rows, so it's OK to do it in SMJ. I'm not sure about streaming join though.

Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:461]: For `KeyWithIndexToValueRowConverterFormatV1`, we don't have this issue because the `covertToValueRow` didn't change the `currentValue`?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:461]: Yes. KeyWithIndexToValueRowConverterFormatV1 simply uses the row state store will provide, which is expected to be unique instance (as we copy it when putting to state store).

I've had time to go through callers, and I'm still hesitated to let callers deal with copy, because the row to take care is inside of the ValueAndMatchPair, and some of current callers are even re-packaging it to other class, e.g. KeyWithIndexAndValue for getAll(), so not only direct callers should take care of it but caller of the caller may have to take care of it. It would bring too much complication.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:461]: I recall in Spark SQL, there are some places similar and usually we make it as the caller's responsibility to call copy. Is this a code path critical for performance?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala]: Looks like we'd prefer to open maximizing performance for all callers. I changed to not copy whenever possible, with adding NOTE in method comments on all impacting methods. be34258

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala]: The disadvantage for the approach is, given the contract is applied on the interface level, row from format V1 now would get unnecessary copy whereas it doesn't need to. It only applies to the evicted rows and ideally end users would need to go with format V2, hope it is OK.

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala [hunk 3]
```diff
@@ -259,6 +269,9 @@ class SymmetricHashJoinStateManager(
           return null
         }
 
+        // Make a copy on value row, as below cleanup logic may update the value row silently.
+        currentValue = currentValue.copy(value = currentValue.value.copy())
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:273]: so this is the only place to do copy?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:273]: Yes. That wasn't necessary for format V1 as the original row was stored into state store, and state store (strictly saying, the implementation of HDFS state store provider) makes sure these rows are copied version.

For other places, it can propagate to the callers outside of state manager, and looks like these callers don't need to copy the row. (It's super tricky for me to determine whether the copy is necessary or not, if the code is not in a simple loop or stream.)

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:273]: After seeing the new changes, I think the first version looks better. The caller sides is nested and we still have necessary copies for v1 format. What do you think? @viirya

Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:273]: Yep, also prefer the first approach personally. As the issue was in v2 format, the first version is a straightforward way. 
@cloud-fan typo? `... unnecessary copies for v1 format`

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by dongjoon-hyun** [comment]
https://github.com/apache/spark/pull/29017 is created based on @HeartSaVioR 's reporting.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by xuanyuanking** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala (hunk 3 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:273]: OK I'll roll back the change. I'll also leave a commit sha so we can do back and forth depending on the decision.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala]: Just reverted the latest commit to leave the history and pick the commit selectively according to the decision.

**Comment by HeartSaVioR** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:461]: we need to update it. It's not because of storing, but the caller side updates it.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:461]: I'm not sure I get it. The problem occurs when caller reads the value lately and there's "another" interaction with the method in the middle of. I agree the sentence is not clear as well though.

Would it be better if we can rephrase as "... during execution, so that caller can safely read the value in any time" ?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala:461]: SGTM. I was referring to https://github.com/apache/spark/pull/28975#discussion_r450016555

**Comment by xuanyuanking** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala [hunk 4]
```diff
@@ -996,4 +997,47 @@ class StreamingOuterJoinSuite extends StreamTest with StateStoreMetricsTest with
       )
     }
   }
+
+  test("SPARK-32148 stream-stream join regression on Spark 3.0.0") {
+    val input1 = MemoryStream[(Timestamp, String, String)]
+    val df1 = input1.toDF
+      .selectExpr("_1 as eventTime", "_2 as id", "_3 as comment")
```
Comment (gaborgsomogyi) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:1004]: Any specific reason why not use `select`? I don't see any expression here.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala [hunk 5]
```diff
@@ -996,4 +997,47 @@ class StreamingOuterJoinSuite extends StreamTest with StateStoreMetricsTest with
...
+      .withWatermark(s"eventTime", "2 minutes")
+
+    val input2 = MemoryStream[(Timestamp, String, String)]
+    val df2 = input2.toDF
+      .selectExpr("_1 as eventTime", "_2 as id", "_3 as name")
```
Comment (gaborgsomogyi) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:1009]: Same here.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala [hunk 6]
```diff
@@ -996,4 +997,47 @@ class StreamingOuterJoinSuite extends StreamTest with StateStoreMetricsTest with
...
+      .withWatermark(s"eventTime", "4 minutes")
+
+    val joined = df1.as("left")
+      .join(df2.as("right"),
+        expr(s"""
+                |left.id = right.id AND left.eventTime BETWEEN
```
Comment (gaborgsomogyi) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:1015]: Nit: indent

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala [hunk 7]
```diff
@@ -996,4 +997,47 @@ class StreamingOuterJoinSuite extends StreamTest with StateStoreMetricsTest with
...
+      .withWatermark(s"eventTime", "2 minutes")
+
+    val input2 = MemoryStream[(Timestamp, String, String)]
+    val df2 = input2.toDF
+      .selectExpr("_1 as eventTime", "_2 as id", "_3 as name")
+      .withWatermark(s"eventTime", "4 minutes")
+
+    val joined = df1.as("left")
+      .join(df2.as("right"),
+        expr(s"""
```
Comment (gaborgsomogyi) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:1014]: Why string interpolation needed?

**Review by gaborgsomogyi — COMMENTED** [review]
The change itself looks good, some minors found.

**Comment by gaborgsomogyi** [comment]
I'm also suffering from flaky executions, hope this round will pass.

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala (hunk 4 above)
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:1004]: I guess it's pretty much simpler and more readable than `select('_1.as("eventTime"), '_2.as("id"), '_3.as("comment"))` (or even with col(...) if ' notation doesn't work for _1, _2, _3).

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala (hunk 5 above)
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:1009]: Same here as well.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala (hunk 7 above)
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:1014]: Ah that's not necessary. Will remove.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala (hunk 6 above)
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala]: The indentation of """ looks vary on the codebase, and I can find same indentation on the codebase.

**Review by xuanyuanking — APPROVED** [review]
LGTM

**Review by gaborgsomogyi — APPROVED** [review]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!
